neo/NeoSystem.cs

```diff
@@ -29,18 +42,40 @@ public class NeoSystem : IDisposable
         public IActorRef Consensus { get; private set; }
         public RpcServer RpcServer { get; private set; }
 
+        private class ShutdownReason : CoordinatedShutdown.Reason
+        {
+            public static CoordinatedShutdown.Reason Instance = new ShutdownReason();
+
+            private ShutdownReason()
+            {
+            }
+        }
+
+        private async Task<Done> StopBlockchainAndWaitForStopped()
+        {
+            ActorSystem.Stop(Blockchain);
+            await Neo.Ledger.Blockchain.Singleton.WaitForStopped();
+            return Done.Instance;
+        }
+
         public NeoSystem(Store store)
         {
             this.Blockchain = ActorSystem.ActorOf(Ledger.Blockchain.Props(this, store));
             this.LocalNode = ActorSystem.ActorOf(Network.P2P.LocalNode.Props(this));
             this.TaskManager = ActorSystem.ActorOf(Network.P2P.TaskManager.Props(this));
             Plugin.LoadPlugins(this);
+            // NOTE: The user can add additional tasks to these shutdown phases to further delay shutdown.
+            CoordinatedShutdown.Get(ActorSystem).AddTask("wait-for-localnode-shutdown", "wait-for-localnode-stopped",
+                Neo.Network.P2P.LocalNode.Singleton.WaitForStopped);
+            CoordinatedShutdown.Get(ActorSystem).AddTask("wait-for-neo-shutdown", "wait-for-blockchain-stopped",
+                StopBlockchainAndWaitForStopped);
         }
 
         public void Dispose()
```

Comment:
I found a easier way to dispose `NeoSystem`:

```c#
public void Dispose()
{
    RpcServer?.Dispose();
    ActorSystem.Terminate();
    ActorSystem.WhenTerminated.Wait();
    ActorSystem.Dispose();
}
```

Reply:
```
ActorSystem.Terminate();
ActorSystem.WhenTerminated.Wait();
```
It seems doing this before dispose is still actually needed. I added those changes to this PR, but still want to do a bit more checking if my other changes are all needed for use cases.

Reply:
```
ActorSystem.Terminate();
ActorSystem.WhenTerminated.Wait();
```
It seems doing this before dispose is still actually needed. I added those changes to this PR, but still want to do a bit more checking if my other changes are all needed for use cases.

Reply:
Actually, by default the `CoordinatedShutdown` runs `Terminate()` and waiting for coordinated shutdown waits until termination completes. So Those are only needed when not using `CoordinatedShutdown`. I think we want to keep `CoordinatedShutdown though so we can ensure order of stopping the actors on shutdown. 

If I create another actor upon shutdown to watch the death of the other actors, I can avoid using semaphores to determine when the actors have actually stopped, which seems kind of hacky. I'm working on adjusting for this now.

Reply:
I've now adjusted to avoid adding the Semaphores into the actors themselves.